routes/util.py

```diff
@@ -250,21 +254,23 @@ def url_for(*args, **kargs):
             newargs = _screenargs(kargs, config.mapper, environ)
         anchor = newargs.pop('_anchor', None) or anchor
         host = newargs.pop('_host', None) or host
-        protocol = newargs.pop('_protocol', None) or protocol
+        protocol = newargs.pop('_protocol', protocol)
         url = config.mapper.generate(*route_args, **newargs)
     if anchor is not None:
         url += '#' + _url_quote(anchor, encoding)
-    if host or protocol or qualified:
+    if host or protocol is not None or qualified:
```

Comment:
Can you paren in the (protocol is not None) bit? I've had problems in the past where order of operation binding messes things up in non-obvious ways.